XBOXOLD.cpp

```diff
@@ -233,37 +250,108 @@ uint8_t XBOXOLD::Init(uint8_t parent, uint8_t port, bool lowspeed) {
+#endif
+}
+
 /* Performs a cleanup after failed Init() attempt */
 uint8_t XBOXOLD::Release() {
         XboxConnected = false;
         pUsb->GetAddressPool().FreeAddress(bAddress);
         bAddress = 0;
+        bNumEP = 1; // Must have to be reset to 1
+        qNextPollTime = 0; // Reset next poll time
+        pollInterval = 0;
         bPollEnable = false;
+#ifdef DEBUG_USB_HOST
+        Notify(PSTR("\r\nXbox Controller Disconnected\r\n"), 0x80);
+#endif
         return 0;
 }
 
 uint8_t XBOXOLD::Poll() {
+        uint8_t rcode = 0;
+
         if(!bPollEnable)
                 return 0;
-        uint16_t BUFFER_SIZE = EP_MAXPKTSIZE;
-        pUsb->inTransfer(bAddress, epInfo[ XBOX_INPUT_PIPE ].epAddr, &BUFFER_SIZE, readBuf); // input on endpoint 1
-        readReport();
-#ifdef PRINTREPORT
-        printReport(BUFFER_SIZE); // Uncomment "#define PRINTREPORT" to print the report send by the Xbox controller
+
+        if((int32_t)((uint32_t)millis() - qNextPollTime) >= 0L) { // Do not poll if shorter than polling interval
+                qNextPollTime = (uint32_t)millis() + pollInterval; // Set new poll time
+                uint16_t length =  (uint16_t)epInfo[ XBOX_INPUT_PIPE ].maxPktSize; // Read the maximum packet size from the endpoint
+                uint8_t rcode = pUsb->inTransfer(bAddress, epInfo[ XBOX_INPUT_PIPE ].epAddr, &length, readBuf, pollInterval);
+                if(!rcode) {
+                        readReport();
+#ifdef PRINTREPORT // Uncomment "#define PRINTREPORT" to print the report send by the Xbox ONE Controller
+                        for(uint8_t i = 0; i < length; i++) {
+                                D_PrintHex<uint8_t > (readBuf[i], 0x80);
+                                Notify(PSTR(" "), 0x80);
+                        }
+                        Notify(PSTR("\r\n"), 0x80);
 #endif
-        return 0;
+                }
+#ifdef DEBUG_USB_HOST
+                else if(rcode != hrNAK) { // Not a matter of no update to send
+                        Notify(PSTR("\r\nXbox Poll Failed, error code: "), 0x80);
+                        NotifyFail(rcode);
+                }
+#endif
+        }
+        return rcode;
 }
 
 void XBOXOLD::readReport() {
         ButtonState = readBuf[2];
+        void *p = readBuf;
 
         for(uint8_t i = 0; i < sizeof (buttonValues); i++)
                 buttonValues[i] = readBuf[i + 4]; // A, B, X, Y, BLACK, WHITE, L1, and R1
 
-        hatValue[LeftHatX] = (int16_t)(((uint16_t)readBuf[12] << 8) | readBuf[13]);
-        hatValue[LeftHatY] = (int16_t)(((uint16_t)readBuf[14] << 8) | readBuf[15]);
-        hatValue[RightHatX] = (int16_t)(((uint16_t)readBuf[16] << 8) | readBuf[17]);
-        hatValue[RightHatY] = (int16_t)(((uint16_t)readBuf[18] << 8) | readBuf[19]);
+        hatValue[LeftHatX] = *(int16_t *)(p + 12);
+        hatValue[LeftHatY] = *(int16_t *)(p + 14);
+        hatValue[RightHatX] = *(int16_t *)(p + 16);
+        hatValue[RightHatY] = *(int16_t *)(p + 18);
```

Comment:
I guess the bug here was the data is little-endian.

Please use shifts, as before, so this also works on big-endian MCUs.

Reply:
Also this does not compile: https://github.com/felis/USB_Host_Shield_2.0/actions/runs/9931283871/job/27451495122?pr=818#step:7:315

Reply:
Yes, it looks like I was too accustomed to gcc and others enabling by default void pointer arithmetic.
I'll rework it using shifts and ORs. Thanks.

Reply:
Fixed.